Introduction of support id

LyzeOfKiel opened the pull request:
Changes:
* Add `traceId` support. Trace ID can be any string, including empty string. If any additional validation is required you can suggest in comments. Now, `/sendTransactions` method accepts only the following requests:
```
{
    transactions: PoolTx[] // same structure as before
    traceId: string
}
```
Closes #100 
* Remove `/sendTransaction` endpoint as it is not used in our client implementation anymore (@EvgenKor maybe we should double-check it).
* Remove `/transactions` endpoint as it is not used in our client implementation anymore.
* Fix spelling for `/limits` response. It also requires changes in client library.

k1rill-fedoseev commented on zp-relayer/router.ts:
```diff
@@ -29,7 +29,6 @@ router.use((err: any, req: Request, res: Response, next: NextFunction) => {
 router.get('/', endpoints.root)
-router.post('/sendTransaction', wrapErr(endpoints.sendTransaction))
 router.post('/sendTransactions', wrapErr(endpoints.sendTransactions))
 router.get('/transactions', wrapErr(endpoints.getTransactions))
```
Should this be removed?

LyzeOfKiel replied:
Yes, this method is deprecated and not used by client library for a while. `/sendTransactions` basically does the same thing but can accept several transactions in one request.

k1rill-fedoseev replied:
I meant this line:
```typescript
router.get('/transactions', wrapErr(endpoints.getTransactions))
```

LyzeOfKiel replied:
Got it. Same thing applies to this endpoint as well. Not used by client any more.

LyzeOfKiel replied:
Will remove it.

k1rill-fedoseev commented on zp-relayer/workers/poolTxWorker.ts:
```diff
@@ -33,27 +33,28 @@ export async function createPoolTxWorker<T extends EstimationType>(
-    const logPrefix = `POOL WORKER: Job ${job.id}:`
-    logger.info('%s processing...', logPrefix)
-    logger.info('Received %s txs', txs.length)
+    const jobLogger = scopedLogger(`POOL WORKER: Job ${job.id}: `, { traceId })
```
I personally would prefer to have all the fields inside of the property object:
```
const workerLogger = logger.child({ worker: 'pool' })
...
const jobLogger = workerLogger.child({ jobId: job.id, traceId })
```
Wdyt?

LyzeOfKiel replied:
Agree, looks more concise. I will update this.

k1rill-fedoseev commented on zp-relayer/workers/poolTxWorker.ts:
```diff
@@ -78,7 +79,7 @@ export async function createPoolTxWorker<T extends EstimationType>(
         await updateNonce(++nonce)
 
-        logger.info(`${logPrefix} TX hash ${txHash}`)
+        jobLogger.info(`TX hash ${txHash}`)
```
I think it makes sense to log all tx-related logs with txHash in the properties object under a`txHash` key:
```
jobLogger.info('Sent tx', { txHash })
```

akolotov reviewed (commented):
Consider to introduce
1/ TrackId for every endpoint so we can monitor activity of a user through all its actions,
2/ ability to turn on/off the requirement to have non-empty trackId through an environment variable

LyzeOfKiel commented:
> Consider to introduce
> 1/ TrackId for every endpoint so we can monitor activity of a user through all its actions,
> 2/ ability to turn on/off the requirement to have non-empty trackId through an environment variable

@akolotov  In this way for `GET` requests `traceId` should be in query args (e.g. `/info?traceId=${traceId})` and for `POST` requests `traceId` should be in request body part?

akolotov commented:
> @akolotov In this way for GET requests traceId should be in query args (e.g. /info?traceId=${traceId}) and for POST requests traceId should be in request body part?

Would it be better to have it as part of HTTP headers? Do you see limitations of this approach?

LyzeOfKiel commented:
> > @akolotov In this way for GET requests traceId should be in query args (e.g. /info?traceId=${traceId}) and for POST requests traceId should be in request body part?
> 
> Would it be better to have it as part of HTTP headers? Do you see limitations of this approach?

Will test this approach.

akolotov commented on zp-relayer/endpoints.ts:
```diff
@@ -9,20 +9,26 @@ import {
 
   const transactions = req.body.transactions as PoolTx[]
-  const traceId = req.body.traceId
+  const traceId = req.headers['trace-id'] as string
```
How about to obfuscate it a bit: `zkbob-support-id`?

LyzeOfKiel commented:
@akolotov Added `trace-id` header for all endpoints except `/`, `/info`, `/params/hash/tree`, `/params/hash/tx`. I thought this would be too redundant. Should we add header validation for them as well? Or maybe we  should remove validation from some other endpoints?
Now, this header is logged for every endpoint if it passes validation. For `/sendTransactions` endpoint header is also logged several times during job processing.

LyzeOfKiel commented:
Also, `RELAYER_REQUIRE_TRACE_ID` env can configure whether header validation is required or not.

EvgenKor commented:
> Also, RELAYER_REQUIRE_TRACE_ID env can configure whether header validation is required or not.

In this way when RELAYER_REQUIRE_TRACE_ID is true we CAN add zkbob-support-id header. If the client didn't set this header, the relayer will no track it. Am I right?

LyzeOfKiel commented:
> > Also, RELAYER_REQUIRE_TRACE_ID env can configure whether header validation is required or not.
> 
> In this way when RELAYER_REQUIRE_TRACE_ID is true we CAN add zkbob-support-id header. If the client didn't set this header, the relayer will no track it. Am I right?

If `RELAYER_REQUIRE_TRACE_ID` is `true` then relayer will require trace id header to be present, otherwise request will be rejected with code 400.
In other case, relayer will not validate the header. So, it doesn't matter whether it is present or not.

k1rill-fedoseev commented on zp-relayer/validation/validation.ts:
```diff
@@ -77,39 +79,21 @@ const AjvSendTransactionSchema: JSONSchemaType<PoolTx> = {
-  offset: number
-  optimistic: boolean
-}> = {
+const AjvSendTransactionsSchema: JSONSchemaType<{ transactions: PoolTx[] }> = {
```
Should we really change the schema here?

LyzeOfKiel replied:
You mean that if don't use trace id as request body property any more then we can restore previously used function signature?

k1rill-fedoseev replied:
Yeah

LyzeOfKiel replied:
Agree, will restore previous signature

LyzeOfKiel commented:
Updates:
* Now, `/sendTransactions` endpoint accepts only an array of transaction as before.
* Changed `trace-id` to `zkbob-support-id`.
* Refactored validation logic so that validation errors are now implicitly thrown and handled by express middleware.

k1rill-fedoseev commented on zp-relayer/router.ts:
```diff
@@ -29,6 +31,13 @@ router.use((err: any, req: Request, res: Response, next: NextFunction) => {
 
+router.use((req: Request, res: Response, next: NextFunction) => {
+  if (config.requireTraceId && req.headers[TRACE_ID]) {
+    logger.info('TraceId %s %s', req.headers[TRACE_ID], req.path)
```
Let's put traceId and request path both under respective json keys in the log.

Http request from users are currently logged to a separate file, maybe it makes sense to put traceId there and log these messages to console as well.

LyzeOfKiel replied:
>Let's put traceId and request path both under respective json keys in the log.

Agree

> maybe it makes sense to put traceId there and log these messages to console as well.

`zp.log` where every request is dumped, already includes all headers, so trace id will be saved there as well.
I don't think logging everything to console that is saved in `zp.log` is a good idea, as logs will be mostly flooded with such request logs which are >500 characters each

k1rill-fedoseev replied:
> I don't think logging everything to console that is saved in zp.log is a good idea, as logs will be mostly flooded with such request logs which are >500 characters each

I think we do not need all of these 500 characters, but ok, let's leave it for future discussion/PR anyway.

akolotov commented:
Will be merged as soon as https://github.com/zkBob/zeropool-relayer/pull/116 is closed.

akolotov commented:
@LyzeOfKiel one more thing - please update the README file with new information about the request header.

akolotov commented:
@LyzeOfKiel sorry but another merge conflict appeared
